File: src/domain/ProgressCalculationEngine.ts
import { Routine, RoutineFrequency, RoutineLog, DailyProgressResult, Goal, GoalProgressResult } from '../types';

export class ProgressCalculationEngine {
  /**
   * Parses comma-separated day numbers (1=Monday .. 7=Sunday).
   */
  static parseFrequencyDays(daysStr: string): Set<number> {
    if (!daysStr || daysStr.trim() === '') {
      return new Set([1, 2, 3, 4, 5, 6, 7]);
    }
    const days = new Set<number>();
    daysStr.split(',').forEach((part) => {
      const parsed = parseInt(part.trim(), 10);
      if (!isNaN(parsed) && parsed >= 1 && parsed <= 7) {
        days.add(parsed);
      }
    });
    return days.size > 0 ? days : new Set([1, 2, 3, 4, 5, 6, 7]);
  }

  /**
   * Returns standard 1 (Monday) to 7 (Sunday) for a given YYYY-MM-DD string.
   */
  static getDayOfWeek(dateStr: string): number {
    const parts = dateStr.split('-').map(Number);
    const date = new Date(parts[0], parts[1] - 1, parts[2]);
    const jsDay = date.getDay(); // 0 is Sunday, 1 is Monday ... 6 is Saturday
    return jsDay === 0 ? 7 : jsDay;
  }

  /**
   * Formats a Date object to YYYY-MM-DD
   */
  static formatDate(date: Date): string {
    const y = date.getFullYear();
    const m = String(date.getMonth() + 1).padStart(2, '0');
    const d = String(date.getDate()).padStart(2, '0');
    return `${y}-${m}-${d}`;
  }

  /**
   * Returns today's date in YYYY-MM-DD format
   */
  static getTodayStr(): string {
    return this.formatDate(new Date());
  }

  /**
   * Compares two YYYY-MM-DD date strings: -1 if a < b, 0 if a == b, 1 if a > b
   */
  static compareDates(a: string, b: string): number {
    return a.localeCompare(b);
  }

  /**
   * Adds or subtracts days to a YYYY-MM-DD string
   */
  static addDays(dateStr: string, days: number): string {
    const parts = dateStr.split('-').map(Number);
    const date = new Date(parts[0], parts[1] - 1, parts[2]);
    date.setDate(date.getDate() + days);
    return this.formatDate(date);
  }

  /**
   * Checks if a routine is naturally scheduled to occur on the specified date.
   */
  static isRoutineActiveOnDate(routine: Routine, dateStr: string): boolean {
    if (routine.isPaused) return false;

    // Check routine start date
    if (routine.startDate && this.compareDates(dateStr, routine.startDate) < 0) {
      return false;
    }

    // Check routine end date if set
    if (routine.endDate && this.compareDates(dateStr, routine.endDate) > 0) {
      return false;
    }

    const dayOfWeek = this.getDayOfWeek(dateStr);

    switch (routine.frequency) {
      case RoutineFrequency.DAILY:
        return true;
      case RoutineFrequency.WEEKDAYS:
        return dayOfWeek >= 1 && dayOfWeek <= 5;
      case RoutineFrequency.WEEKENDS:
        return dayOfWeek === 6 || dayOfWeek === 7;
      case RoutineFrequency.CUSTOM_DAYS:
        return this.parseFrequencyDays(routine.frequencyDays).has(dayOfWeek);
      case RoutineFrequency.WEEKLY:
        // By default first day of parsed frequencyDays, or Monday
        const targetDays = this.parseFrequencyDays(routine.frequencyDays);
        return targetDays.has(dayOfWeek);
      default:
        return true;
    }
  }

  /**
   * Calculates comprehensive daily progress for a given date.
   */
  static calculateDailyProgress(
    date: string,
    routines: Routine[],
    logs: RoutineLog[]
  ): DailyProgressResult {
    // 1. Identify routines scheduled on this date
    // Note: also check logs to see if any routine was rescheduled AWAY from this date or TO this date
    const routineMap = new Map<number, Routine>();
    routines.forEach((r) => routineMap.set(r.id, r));

    // Logs specifically tied to this date
    const dateLogs = logs.filter((log) => log.date === date);
    const dateLogMap: Record<number, RoutineLog> = {};
    dateLogs.forEach((log) => {
      dateLogMap[log.routineId] = log;
    });

    // Also look for logs on other dates that rescheduled a routine TO this date
    const rescheduledToThisDateRoutineIds = new Set<number>();
    logs.forEach((log) => {
      if (log.rescheduledToDate === date) {
        rescheduledToThisDateRoutineIds.add(log.routineId);
      }
    });

    const plannedList: Routine[] = [];

    routines.forEach((routine) => {
      const logOnDate = dateLogMap[routine.id];
      // If rescheduled away from this date, don't include in today's active requirements
      if (logOnDate && logOnDate.rescheduledToDate && logOnDate.rescheduledToDate !== date) {
        return;
      }

      const naturallyActive = this.isRoutineActiveOnDate(routine, date);
      const rescheduledIn = rescheduledToThisDateRoutineIds.has(routine.id);

      if (naturallyActive || rescheduledIn) {
        plannedList.push(routine);
      }
    });

    // Sort planned routines by scheduled time
    plannedList.sort((a, b) => {
      if (a.timeHour !== b.timeHour) return a.timeHour - b.timeHour;
      return a.timeMinute - b.timeMinute;
    });

    const completedRoutineIds = new Set<number>();
    const skippedRoutineIds = new Set<number>();

    plannedList.forEach((r) => {
      const log = dateLogMap[r.id];
      if (log) {
        if (log.isCompleted) {
          completedRoutineIds.add(r.id);
        } else if (log.isSkipped) {
          skippedRoutineIds.add(r.id);
        }
      }
    });

    const totalPlanned = plannedList.length;
    const completedCount = completedRoutineIds.size;
    const skippedCount = skippedRoutineIds.size;
    const completionPercentage = totalPlanned > 0
      ? Math.min(100, Math.round((completedCount / totalPlanned) * 1000) / 10)
      : 0.0;

    return {
      date,
      totalPlanned,
      completedCount,
      skippedCount,
      completionPercentage,
      plannedRoutines: plannedList,
      completedRoutineIds,
      skippedRoutineIds,
      routineLogMap: dateLogMap,
    };
  }

  /**
   * Calculates progress for a specific goal based on linked routines and manual currentValue.
   */
  static calculateGoalProgress(
    goal: Goal,
    routines: Routine[],
    logs: RoutineLog[]
  ): GoalProgressResult {
    const linkedRoutines = routines.filter((r) => r.linkedGoalId === goal.id);
    const linkedRoutineIds = linkedRoutines.map((r) => r.id);

    let completedRoutinesCount = 0;
    if (linkedRoutineIds.length > 0) {
      const linkedIdSet = new Set(linkedRoutineIds);
      logs.forEach((l) => {
        if (linkedIdSet.has(l.routineId) && l.isCompleted) {
          completedRoutinesCount++;
        }
      });
    }

    // Determine current value: prioritize manual progress if greater than 0, otherwise completed count
    const effectiveCurrent = Math.max(goal.currentValue, completedRoutinesCount);
    const target = goal.targetValue > 0 ? goal.targetValue : 1;
    const percentage = Math.min(100, Math.round((effectiveCurrent / target) * 1000) / 10);
    const isCompleted = percentage >= 100;

    return {
      goalId: goal.id,
      currentValue: effectiveCurrent,
      targetValue: goal.targetValue,
      percentage,
      completedRoutinesCount,
      totalLinkedRoutines: linkedRoutines.length,
      linkedRoutineIds,
      isCompleted,
    };
  }
}

File: app/src/main/java/com/example/domain/ProgressCalculationEngine.kt
package com.example.domain

import com.example.data.model.*
import java.time.DayOfWeek
import java.time.LocalDate
import java.time.format.DateTimeFormatter

data class DailyProgressResult(
    val date: LocalDate,
    val totalPlanned: Int,
    val completedCount: Int,
    val skippedCount: Int,
    val missedCount: Int,
    val completionPercentage: Double?, // null if totalPlanned == 0
    val plannedRoutines: List<Routine>,
    val completedRoutineIds: Set<Long>,
    val skippedRoutineIds: Set<Long>,
    val routineLogMap: Map<Long, RoutineLog>
)

data class GoalProgressResult(
    val goalId: Long,
    val currentValue: Double,
    val targetValue: Double,
    val percentage: Double, // 0.0 to 100.0
    val isCompleted: Boolean,
    val linkedRoutineCount: Int,
    val milestones: List<GoalMilestone>,
    val newlyAchievedMilestones: List<GoalMilestone>
)

object ProgressCalculationEngine {
    val DATE_FORMATTER: DateTimeFormatter = DateTimeFormatter.ISO_LOCAL_DATE // YYYY-MM-DD

    /**
     * Checks if a routine is planned to occur on a given date.
     */
    fun isRoutineScheduledForDate(
        routine: Routine,
        date: LocalDate,
        logsForRoutine: List<RoutineLog> = emptyList()
    ): Boolean {
        // Paused routines do not generate expected occurrences
        if (routine.isPaused) return false

        val dateStr = date.format(DATE_FORMATTER)

        // Check if start date is reached
        val startDate = try {
            LocalDate.parse(routine.startDate, DATE_FORMATTER)
        } catch (e: Exception) {
            LocalDate.MIN
        }
        if (date.isBefore(startDate)) return false

        // Check if end date has passed
        if (!routine.endDate.isNullOrBlank()) {
            val endDate = try {
                LocalDate.parse(routine.endDate, DATE_FORMATTER)
            } catch (e: Exception) {
                LocalDate.MAX
            }
            if (date.isAfter(endDate)) return false
        }

        // Check if this routine was explicitly rescheduled AWAY from this date
        val logForDate = logsForRoutine.find { it.date == dateStr }
        if (logForDate?.status == LogStatus.RESCHEDULED) {
            return false
        }

        // Check if another occurrence was rescheduled TO this date
        val rescheduledToToday = logsForRoutine.any {
            it.status == LogStatus.RESCHEDULED && it.rescheduledToDate == dateStr
        }
        if (rescheduledToToday) {
            return true
        }

        // Check recurring frequency
        return when (routine.frequency) {
            RoutineFrequency.DAILY -> true
            RoutineFrequency.WEEKDAYS -> {
                val dow = date.dayOfWeek
                dow != DayOfWeek.SATURDAY && dow != DayOfWeek.SUNDAY
            }
            RoutineFrequency.WEEKLY_X_TIMES -> {
                // For X times per week routines, they are planned if days are configured or open
                val activeDays = parseFrequencyDays(routine.frequencyDays)
                if (activeDays.isEmpty()) true else activeDays.contains(date.dayOfWeek.value)
            }
            RoutineFrequency.CUSTOM_DAYS -> {
                val activeDays = parseFrequencyDays(routine.frequencyDays)
                activeDays.contains(date.dayOfWeek.value)
            }
        }
    }

    fun parseFrequencyDays(daysStr: String): Set<Int> {
        if (daysStr.isBlank()) return setOf(1, 2, 3, 4, 5, 6, 7)
        return daysStr.split(",")
            .mapNotNull { it.trim().toIntOrNull() }
            .filter { it in 1..7 }
            .toSet()
    }

    /**
     * Calculates daily progress metrics for a given date.
     */
    fun calculateDailyProgress(
        routines: List<Routine>,
        logsForDate: List<RoutineLog>,
        allLogs: List<RoutineLog> = emptyList(),
        date: LocalDate
    ): DailyProgressResult {
        val dateStr = date.format(DATE_FORMATTER)
        val logsByRoutine = allLogs.groupBy { it.routineId }

        val planned = routines.filter { routine ->
            isRoutineScheduledForDate(
                routine = routine,
                date = date,
                logsForRoutine = logsByRoutine[routine.id] ?: emptyList()
            )
        }

        val logMap = mutableMapOf<Long, RoutineLog>()
        logsForDate.forEach { log ->
            logMap[log.routineId] = log
        }

        val completedIds = mutableSetOf<Long>()
        val skippedIds = mutableSetOf<Long>()

        for (routine in planned) {
            val log = logMap[routine.id]
            if (log != null) {
                when (log.status) {
                    LogStatus.COMPLETED -> completedIds.add(routine.id)
                    LogStatus.SKIPPED -> skippedIds.add(routine.id)
                    LogStatus.RESCHEDULED -> { /* Not planned on this day */ }
                }
            }
        }

        val completedCount = completedIds.size
        val skippedCount = skippedIds.size
        val totalPlanned = planned.size

        // In terms of required work:
        // If all tasks are skipped, it's a neutral day (100% skipped)
        // Qualifying completion % = completed / (totalPlanned - skipped) or completed / totalPlanned
        val activePlanned = (totalPlanned - skippedCount).coerceAtLeast(0)
        val missedCount = (activePlanned - completedCount).coerceAtLeast(0)

        val percentage = when {
            totalPlanned == 0 -> null // No scheduled routines
            activePlanned == 0 -> 100.0 // All scheduled were legitimately skipped
            else -> {
                val raw = (completedCount.toDouble() / activePlanned.toDouble()) * 100.0
                (Math.round(raw * 10.0) / 10.0).coerceIn(0.0, 100.0)
            }
        }

        return DailyProgressResult(
            date = date,
            totalPlanned = totalPlanned,
            completedCount = completedCount,
            skippedCount = skippedCount,
            missedCount = missedCount,
            completionPercentage = percentage,
            plannedRoutines = planned,
            completedRoutineIds = completedIds,
            skippedRoutineIds = skippedIds,
            routineLogMap = logMap
        )
    }

    /**
     * Calculates the progress of a Goal based on measurable targets and linked routines.
     */
    fun calculateGoalProgress(
        goal: Goal,
        linkedRoutines: List<Routine>,
        allLogs: List<RoutineLog>,
        milestones: List<GoalMilestone>
    ): GoalProgressResult {
        val linkedRoutineIds = linkedRoutines.map { it.id }.toSet()
        val relevantLogs = allLogs.filter { it.routineId in linkedRoutineIds && it.status == LogStatus.COMPLETED }

        var calculatedCurrent = 0.0

        if (goal.targetValue > 0) {
            val hasMeasurableRoutines = linkedRoutines.any {
                it.taskType != TaskType.CHECKBOX
            }

            if (hasMeasurableRoutines) {
                // Sum actual values (duration, quantity, count)
                calculatedCurrent = relevantLogs.sumOf { it.actualValue }
            } else {
                // Checkbox or frequency count
                calculatedCurrent = relevantLogs.size.toDouble()
            }
        }

        val target = if (goal.targetValue > 0) goal.targetValue else 1.0
        val rawPercentage = (calculatedCurrent / target) * 100.0
        val percentage = (Math.round(rawPercentage * 10.0) / 10.0).coerceIn(0.0, 100.0)
        val isCompleted = calculatedCurrent >= target || goal.status == GoalStatus.COMPLETED

        // Check milestones
        val newlyAchieved = mutableListOf<GoalMilestone>()
        val updatedMilestones = milestones.map { milestone ->
            if (milestone.achievedAt == null && percentage >= milestone.targetPercentage) {
                val achieved = milestone.copy(achievedAt = System.currentTimeMillis())
                newlyAchieved.add(achieved)
                achieved
            } else {
                milestone
            }
        }

        return GoalProgressResult(
            goalId = goal.id,
            currentValue = Math.round(calculatedCurrent * 100.0) / 100.0,
            targetValue = goal.targetValue,
            percentage = percentage,
            isCompleted = isCompleted,
            linkedRoutineCount = linkedRoutines.size,
            milestones = updatedMilestones,
            newlyAchievedMilestones = newlyAchieved
        )
    }
}
